ng-dev/auth/shared/oauth.ts

```diff
@@ -0,0 +1,244 @@
+  scope,
+}: OAuthDanceConfig): Promise<TokenResponse> {
+  if (client_id === undefined) {
+    throw Error();
+  }
+
+  // Set up and configure the authentication url to initiate the OAuth dance.
+  const url = new URL(deviceAuthEndpoint);
+  url.searchParams.append('scope', scope);
+  url.searchParams.append('client_id', client_id);
+  url.searchParams.append('response_type', AuthorizationRequest.RESPONSE_TYPE_CODE);
+
+  // Request the device and user codes to begin OAuth dance for our application.
+  // https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-1:-request-device-and-user-codes
+  const response = await fetch(url.toString(), {
+    method: 'POST',
+    headers: {'Accept': 'application/json'},
+  }).then(
+    (resp) =>
+      resp.json() as unknown as {
+        verification_uri: string;
+        verification_url: string;
+        interval: number;
+        user_code: string;
+        expires_in: number;
+        device_code: string;
+      },
+  );
+
+  Log.info(`Please visit: ${response.verification_uri || response.verification_url}`);
+  Log.info(`Enter your one time ID code: ${response.user_code}`);
+
+  /**
+   * The number of milliseconds to add to the requested internal from Google, utilized if Google requests
+   * that our requests slow down.
+   */
+  let pollingBackoff = 2500;
+
+  const oauthDanceTimeout = Date.now() + response.expires_in * 1000;
+
+  while (true) {
+    if (Date.now() > oauthDanceTimeout) {
+      throw {
+        authenticated: false,
+        message: 'Failed to completed OAuth authentication before the user code expired.',
+      };
+    }
+    // Wait for the requested interval before polling, this is done before the request as it is unnecessary to
+    //immediately poll while the user has to perform the auth out of this flow.
+    await new Promise((resolve) => setTimeout(resolve, response.interval * 1000 + pollingBackoff));
+
+    const result = await pollAuthServer(
+      authConfig.tokenEndpoint,
+      response.device_code,
+      client_id,
+      client_secret,
+    );
+    if (!result.error) {
+      return {
+        ...result,
+        idToken: result.id_token,
+        accessToken: result.access_token,
+      };
+    }
+    if (result.error === 'access_denied') {
+      throw {
+        authenticated: false,
+        message: 'Unable to authorize, as access was denied during the OAuth flow.',
+      };
+    }
+
+    if (result.error === 'authorization_pending') {
+      // Update messaging.
+    }
+
+    if (result.error === 'slow_down') {
+      // Update messaging.
+      pollingBackoff += 5000;
+    }
+
+    if (
+      result.error === 'invalid_client' ||
+      result.error === 'unsupported_grant_type' ||
+      result.error === 'invalid_grant' ||
+      result.error === 'invalid_request'
+    ) {
+      throw {
+        authenticated: false,
+        message: result.errorDescription,
+      };
+    }
+  }
+}
+
+async function pollAuthServer(
```

Comment:
nit; this does not implement the polling, but just fetches once. can we rename this function?